platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ImportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,234 @@
+import org.codice.ddf.migration.ImportMigrationContext;
+import org.codice.ddf.migration.ImportMigrationContextProxy;
+import org.codice.ddf.migration.ImportMigrationEntry;
+import org.codice.ddf.migration.MigrationException;
+import org.codice.ddf.migration.MigrationReport;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class extends on the {@link ImportMigrationContext} interface to pre-process exported entries
+ * for configuration objects and compare them with the configuration objects currently in memory.
+ */
+public class ImportMigrationConfigurationAdminContext extends ImportMigrationContextProxy {
+    private static final Logger LOGGER = LoggerFactory.getLogger(
+            ImportMigrationConfigurationAdminContext.class);
+
+    private final ConfigurationAdminMigratable admin;
+
+    private final ConfigurationAdmin configurationAdmin;
+
+    /**
+     * Keeps track of all managed services in memory that were not found in the export file so
+     * we know what to delete at the end.
+     */
+    private final Map<String, Configuration> memoryServices;
+
+    /**
+     * Keeps track of all managed service factories in memory that were not found in the export file
+     * so we know what to delete at the end.
+     */
+    private final Map<String, List<Configuration>> memoryFactoryServices;
+
+    private final Map<String, ImportMigrationConfigurationAdminEntry> exportedServices;
+
+    private final Map<String, List<ImportMigrationConfigurationAdminEntry>> exportedFactoryServices;
+
+    private boolean isValid = true; // until proven otherwise
+
+    public ImportMigrationConfigurationAdminContext(ImportMigrationContext context,
+            ConfigurationAdminMigratable admin, ConfigurationAdmin configurationAdmin,
+            Configuration[] memoryConfigs) {
+        super(context);
+        Validate.notNull(admin, "invalid null configuration admin migratable");
+        Validate.notNull(configurationAdmin, "invalid null configuration admin");
+        Validate.notNull(memoryConfigs, "invalid null configurations");
+        this.admin = admin;
+        this.configurationAdmin = configurationAdmin;
+        // categorize memory configurations
+        this.memoryServices = Stream.of(memoryConfigs)
+                .filter(ConfigurationAdminMigratable::isManagedService)
+                .collect(Collectors.toMap(Configuration::getPid, Function.identity()));
+        this.memoryFactoryServices = Stream.of(memoryConfigs)
+                .filter(ConfigurationAdminMigratable::isManagedServiceFactory)
+                .collect(Collectors.groupingBy(Configuration::getFactoryPid));
+        // categorize exported configurations
+        final ImportMigrationConfigurationAdminEntry[] entries = context.entries()
+                .map(this::proxy)
+                .filter(Objects::nonNull)
```

Comment:
❓ Shouldn't we stop as soon as `isValid` becomes `false` and just clear out all the maps since their content would be useless at that point? This _may_ allow us to get rid of the `if (isValid)` checks in the code, and maybe even the member variable as well.

Reply:
That would prevent us from reporting all possible errors. Remember that we are in the constructor here and I don't want to throw an exception out since we are not aborting the import but just recording errors for it so that is aborted later. The isValid() later is to make sure we return no entries to process. I could have cleared the map before returning from the constructor but since the import will fail soon enough after, I wasn't too concerned with using up that memory until then (which would be used anyway in successful case)